File: 백준/Silver/1874. 스택 수열/스택 수열.py
# silver 2
# 새로운 스택에 집어넣었다 뺐다를 반복하면서 기존에 정의된 stk과 같다면 cal 프린트, 아니면 NO 프린트

from collections import deque
import sys
N = int(sys.stdin.readline())
cnt = 0
stk = []
cal = []
flag = True

for _ in range(N):
    a = int(sys.stdin.readline())

    while cnt < a:
        cnt += 1
        stk.append(cnt)
        cal.append('+')

    if stk[-1] == a:
        stk.pop()
        cal.append('-')
    else:
        flag = False
        break

if flag:
    print(*cal)
else:
    print('NO')

File: 백준/Silver/1874. 스택 수열/스택 수열.cc
// Silver 2. Stack Sequence
// 56ms

#include <iostream>
#include <stack>
#include <queue>

using namespace std;

int n;
stack<int> stk;
queue<char> cal;

int main()
{
    ios_base::sync_with_stdio(0);
    cin.tie(0);

    cin >> n;

    bool flag = 1;

    int cnt = 0;

    stk.push(cnt);

    while (n--)
    {
        int a;

        cin >> a;

        while (cnt < a)
        {
            cnt++;
            stk.push(cnt);
            cal.push('+');
        }

        if (stk.top() == a)
        {
            stk.pop();
            cal.push('-');
        }
        else
        {
            flag = 0;
            break;
        }
    }

    if (flag)
    {
        while (!cal.empty())
        {
            cout << cal.front() << '\n';
            cal.pop();
        }
    }
    else
        cout << "NO";
}